pkg/operator/target_status_test.go

```diff
@@ -0,0 +1,1524 @@
+		Location:              "test-loc",
+		Cluster:               "test-cluster",
+		OperatorNamespace:     "gmp-system",
+		TargetPollConcurrency: 4,
+	}
+	if err := opts.defaultAndValidate(logger); err != nil {
+		t.Fatal("Invalid options:", err)
+	}
+
+	timer := newTestTimer()
+
+	scheme, err := getScheme()
+	if err != nil {
+		t.Fatal("Unable to get scheme")
+	}
+
+	port := int32(19090)
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-a",
+			Namespace: opts.OperatorNamespace,
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Name: "prometheus",
+			}},
+		},
+		Status: corev1.PodStatus{
+			Phase: corev1.PodRunning,
+			PodIP: "127.0.0.1",
+			ContainerStatuses: []corev1.ContainerStatus{{
+				Name:  "prometheus",
+				Ready: true,
+			}},
+		},
+	}
+
+	kubeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(&appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      NameCollector,
+			Namespace: opts.OperatorNamespace,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name: "prometheus",
+						Ports: []corev1.ContainerPort{{
+							Name:          "prom-metrics",
+							ContainerPort: port,
+						}},
+					}},
+				},
+			},
+		},
+	}).WithObjects(&monitoringv1.PodMonitoring{
+		ObjectMeta: metav1.ObjectMeta{Name: "prom-example-1", Namespace: "gmp-test"},
+		Spec: v1.PodMonitoringSpec{
+			Endpoints: []v1.ScrapeEndpoint{{
+				Port: intstr.FromString("metrics"),
+			}},
+		},
+	}).WithObjects(pod).Build()
+
+	prometheusTargetMap := make(map[string]*prometheusv1.TargetsResult, 1)
+	key := getPodKey(pod, port)
+	prometheusTargetMap[key] = &prometheusv1.TargetsResult{
+		Active: []prometheusv1.ActiveTarget{{
+			Health: "up",
+			Labels: map[model.LabelName]model.LabelValue{
+				"instance": model.LabelValue("a"),
+			},
+			ScrapePool:         "PodMonitoring/gmp-test/prom-example-1/metrics",
+			LastError:          "err x",
+			LastScrapeDuration: 1.2,
+		}},
+		Dropped: []prometheusv1.DroppedTarget{},
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+	reconciler := &targetStatusReconciler{
+		ch:           ch,
+		opts:         opts,
+		getTarget:    targetFetchFromMap(prometheusTargetMap),
+		logger:       logger,
+		kubeClient:   kubeClient,
+		timerBuilder: newTestTimerBuilder(timer),
+	}
+
+	expectStatus := func(t *testing.T, description string, expected []monitoringv1.ScrapeEndpointStatus) {
+		// Must poll because status is updated via other thread.
+		var err error
+		if pollErr := wait.Poll(100*time.Millisecond, 2*time.Second, func() (done bool, err error) {
+			var podMonitorings monitoringv1.PodMonitoringList
+			kubeClient.List(ctx, &podMonitorings)
+			switch amount := len(podMonitorings.Items); amount {
+			case 0:
+				err = fmt.Errorf("Could not find %s PodMonitoring", description)
```

Comment:
No, it is necessary because the reconcile happens in a separate goroutine. We have a goroutine that ingests the events and calls reconcile, emulating the native controller runtime event handler.